pkg/operator/certupdater/cert_updater.go

```diff
@@ -0,0 +1,212 @@
+	cu := &CertUpdater{
+		source: source,
+	}
+	return cu
+}
+
+// Start begins polling to periodically update certificates from the available sources.
+func (cu *CertUpdater) Start(ctx context.Context) error {
+	go func() {
+		done := cu.pollingInterval == 0
+		for !done {
+			cu.poll(ctx)
+
+			select {
+			case <-time.After(cu.pollingInterval):
+				// Noop
+			case <-ctx.Done():
+				done = true
+			}
+		}
+	}()
+	return nil
+}
+
+func (cu *CertUpdater) poll(ctx context.Context) error {
+	if cu.source == nil {
+		return fmt.Errorf("certificate source not set")
+	}
+
+	cert, ca, err := cu.source(ctx)
+	if err != nil {
+		return fmt.Errorf("cert source: %w", err)
+	}
+	cu.Lock()
+	cu.currentCert = &cert
+	cu.currentCA = ca
+	cu.Unlock()
+
+	return nil
+}
+
+// GetCA allows access to CA Bundle, if applicable
+func (cu *CertUpdater) GetCA() (*x509.Certificate, error) {
+	cu.RLock()
+	defer cu.RUnlock()
+	return cu.currentCA, nil
+}
+
+// GetCertificate implements tls.Config.GetCertificate. Certificates are updated asynchronously.
+func (cu *CertUpdater) GetCertificate(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	cu.RLock()
+	defer cu.RUnlock()
+	return cu.currentCert, nil
+}
+
+// CertSource defines the common signature of functions that return certificates
+type CertSource func(ctx context.Context) (webhook tls.Certificate, ca *x509.Certificate, err error)
+
+// SourceBase64 sources certificates from base64 strings
+func SourceBase64(certString, keyString string, optionalCAString string) (CertSource, error) {
+	cert, err := base64.StdEncoding.DecodeString(certString)
+	if err != nil {
+		return nil, err
+	}
+	key, err := base64.StdEncoding.DecodeString(keyString)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use cert as CA, if not provided
+	if optionalCAString == "" {
+		return sourcePEM(cert, key, cert)
+	}
+
+	ca, err := base64.StdEncoding.DecodeString(optionalCAString)
+	if err != nil {
+		return nil, err
+	}
+	return sourcePEM(cert, key, ca)
+}
+
+// SourceDir sources certificates from directory on the host
+// Expected Certificate Name: `tls.crt`
+// Expected Private Key Name: `tls.key`
+// Expected CA Certificate Name: `ca.crt` [Optional]
+func SourceDir(dir string) (CertSource, error) {
+	if _, err := os.Stat(dir); err != nil {
+		return nil, err
+	}
+	return SourceFS(os.DirFS(dir).(fs.ReadFileFS))
+}
+
+// SourceFS sources certificates from an `io/fs.FS` abstraction
+// Expected Certificate Name: `tls.crt`
+// Expected Private Key Name: `tls.key`
+// Expected CA Certificate Name: `ca.crt` [Optional]
+func SourceFS(fsys fs.ReadFileFS) (CertSource, error) {
+	certPEM, err := fsys.ReadFile(corev1.TLSCertKey)
+	if err != nil {
+		return nil, err
+	}
+	keyPEM, err := fsys.ReadFile(corev1.TLSPrivateKeyKey)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := tls.X509KeyPair(certPEM, keyPEM)
+	if err != nil {
+		return nil, err
+	}
+
+	caPEM, err := fsys.ReadFile(TLSCAKey)
+	if err != nil {
+		caPEM = certPEM
```

Comment:
The intention is that if a separate CA cert is not provided, then the cert will be used as its own the CA. There is nothing special about generated certificates that makes them different in this regard.

None of the certs we see for the webhook server are going to be signed by a Root CA, because a Kubernetes service address (e.g. `gmp-operator.gmp-system.svc`) is not a public DNS name. So, they're all essentially "self-signed."

Sometimes (read: in GKE) the certs will be signed by the cluster CA (also different from a Root CA). In which case, that should be provided and can be used instead of the cert itself to validate the webhook server cert.